benchmarks/idioms/reduction/reduction.sycl.cxx

```diff
@@ -0,0 +1,80 @@
+// BUILD: add_benchmark(ppm=sycl)
+#include <CL/sycl.hpp>
+#include <rosetta.h>
+#include <vector>
+
+static real kernel(cl::sycl::queue q, cl::sycl::buffer<pbsize_t, 1> resultBuffer, pbsize_t global_size, pbsize_t local_size, pbsize_t n) {
+  pbsize_t sum = 0;
+  pbsize_t length = n;
+  do {
+    q.submit([&](cl::sycl::handler &cgh) {
+      auto resAcc = resultBuffer.get_access<cl::sycl::access::mode::write>(cgh);
+      //cl::sycl::local_accessor<pbsize_t, 1> localSum(cl::sycl::range<1>(local_size), cgh); //Doesn't work with llvm-Intel SYCL
+      cl::sycl::accessor<pbsize_t, 1, cl::sycl::access::mode::read_write, sycl::access::target::local> localSum(cl::sycl::range<1>(local_size), cgh);
+      cgh.parallel_for<class reduction_kernel>(
+          cl::sycl::nd_range<1>(cl::sycl::range<1>(global_size), cl::sycl::range<1>(local_size)),
+          [=](cl::sycl::nd_item<1> item) {
+            idx_t global_id = item.get_global_id(0);
+            idx_t local_id = item.get_local_id(0);
+            idx_t group_size = item.get_local_range(0);
+
+            if (global_id >= length) {
+              localSum[local_id] = 0;
+            } else {
+              localSum[local_id] = resAcc[global_id];
+            }
+            item.barrier();
+
+            for (idx_t s = group_size / 2; s > 0; s /= 2) {
+              if (local_id < s) {
+                localSum[local_id] += localSum[local_id + s];
+              }
+              item.barrier();
+            }
+
+            if (local_id == 0) {
+              resAcc[item.get_group(0)] = localSum[0];
+            }
+          });
+    });
+    q.wait();
+    length = global_size / local_size;
+    global_size = length;
+    if (length % local_size != 0) {
+      global_size = (length / local_size + 1) * local_size;
+    }
+  } while (length > 1);
```

Comment:
Repeated invocation of the kernel is a really inefficient implementation of a reduction. See e.g. http://www.cs.ucr.edu/~nael/217-f15/lectures/217-lec10.pdf for a common implementation techniques call tree reduction.